### ISSUE TITLE
lorrshift1 test failure on Cygwin64

### ISSUE THREAD

**Issue by wbhart** [wbhart/mpir#151]
Simon Sobisch reports that the mpn test for lorrshift1 fails on Cygwin64. 

Config.guess returns penryn-unknown-cygwin

**Comment by wbhart** [comment]
Simon Sobisch reports that removing mpn/x86_64w/core2/lshift1.asm and rshift1.asm results in the test suite passing. It seems that these files are faulty.

**Comment by BrianGladman** [comment]
Since these files work on Visual Studio builds, I don't think the problem lies with them.  It would seem that there is a problem in the integration of YASM and GCC code in the cygwin64 build environment.

**Comment by wbhart** [comment]
They do seem to build ok. They just segfault when run.

Anyway, I'll take a look when I get decent internet access again.

**Comment by wbhart** [comment]
I can confirm this also segfaults on my haswell machine on Cygwin64.

It works fine on msys2 on the same machine. I'm not entirely clear what could be different between the two cases. But I'll try and figure it out.

**Comment by wbhart** [comment]
The files involved are:

mpn/x86_64w/sandybridge/lshift.asm
mpn/x86_64w/sandybridge/rshift.asm

**Comment by wbhart** [comment]
But putting the file in doesn't fix the problem. Which is what one would expect, since Simon's penryn also has the problem, and nothing is missing there.

**Comment by wbhart** [comment]
I've now spent a full day investigating the problem.

The function mpn_lshift1 is assembled, it exists in the library. I can link a small C program against it.

With gdb I was able to determine that the when the call to the function is reached, the call jumps off into an invalid address for mpn_lshift1. In other words, the function is never actually called.

We may have to not support Cygwin64. I don't see how we can make any further progress with the problem.

**Comment by wbhart** [comment]
It seems the initial diagnosis is that the Cygwin64 dynamic linker is truncating the 64 bit address of the functions mpn_lshift1 and mpn_rshift1 to 32 bits. It only seems to do it with these two functions.

JP Flori noted this with Cygwin64 once before with MPN_ZERO (the call is presumably to mpn_store).

However, whilst linking libmpir.la (static library this time), it complains:

libtool: link: warning: undefined symbols not allowed in haswell-unknown-cygwin shared libraries
libtool: link: warning: undefined symbols not allowed in haswell-unknown-cygwin shared libraries

It's curious there's precisely two of these.

**Comment by wbhart** [comment]
The issue turned out to be missing __GMP_DECLSPEC's on mpn_lshift1 and mpn_rshift1 in gmp-impl.h (and a few other functions). These cause dllexport/import calling conventions, which are required on Windows. Without them, it calls the functions in the wrong way.

This problem is now resolved in our github repository master branch. I have checked MPIR now builds and passes all tests on Cygwin 32 and 64, with shared or static libraries.
